### ISSUE TITLE
Hardware interface

### ISSUE THREAD

**Issue by bruno-f-cruz** [neurogears/sarahe-vr-2p#1]
### Camera
- [x]  Camera video acquisition (FLIR)

### Digital Input
- [x] Camera Sync pulse (efference copy)
- [x] ScanImage sync pulse
- [x] Screen Photodiode
- [x] Solenoid Valve (efference copy) 

### NI-DAQ Analog Input
- [x] Rotary encoder (Implemented via CounterInput)
- [x] Piezo lick detector
- [x] Photostimulation (efference copy) 

### Digital Output
- [x] Solenoid Valve
- [x] Camera Sync Pulse (implemented via CounterOutput)

### Analog Output
- [x] Photostimulation LED waveform

**Comment by bruno-f-cruz** [comment]
Here's a small draft for the counter output to be tested.

```c#

using System;
using System.Reactive.Linq;
using OpenCV.Net;
using NationalInstruments.DAQmx;
using System.Reactive.Disposables;
using System.ComponentModel;
using System.Reactive;

namespace Bonsai.DAQmx
{
    /// <summary>
    /// Represents an operator that generates voltage signals in one or more DAQmx analog
    /// output channels from a sequence of sample buffers.
    /// </summary>
    [DefaultProperty("Channel")]
    [Description("Generates voltage signals in one or more DAQmx analog output channels from a sequence of sample buffers.")]
    public class CounterOutput : Source<Unit>
    {
        private string counterChannel = string.Empty;
        public string CounterChannel
        {
            get { return counterChannel; }
            set { counterChannel = value; }
        }

        private string triggerSource = string.Empty;
        public string TriggerSource
        {
            get { return triggerSource; }
            set { triggerSource = value; }
        }

        private double frequency = 1;
        public double Frequency
        {
            get { return frequency; }
            set { frequency = value; }
        }

        private double dutyCycle = 0.5;
        public double DutyCycle
        {
            get { return dutyCycle; }
            set { dutyCycle = value; }
        }

        private int? numberOfSamples = null;
        public int? NumberOfSamples
        {
            get { return numberOfSamples; }
            set { numberOfSamples = value; }
        }

        public override IObservable<Unit> Generate()
        {
            return Observable.Using(
                () => new Task(),
                task => Observable.Create<Unit>(observer =>
                {
                    task.COChannels.CreatePulseChannelFrequency(
                        counterChannel,
                        "",
                        COPulseFrequencyUnits.Hertz,
                        COPulseIdleState.Low,
                        0, frequency, dutyCycle);
                    if (numberOfSamples.HasValue)
                    {
                        task.Timing.ConfigureImplicit(SampleQuantityMode.FiniteSamples, numberOfSamples.Value);
                    }
                    else
                    {
                        task.Timing.ConfigureImplicit(SampleQuantityMode.ContinuousSamples);
                    }
                    task.Control(TaskAction.Verify);
                    task.Control(TaskAction.Commit);
                    task.Start();

                    task.Done += (sender, e) =>
                    {
                        if (e.Error != null) observer.OnError(e.Error);
                        else observer.OnCompleted();
                    };
                    return Disposable.Create(task.Stop);
                }
            )
            );
        }
    }
}
```

**Comment by selnozahy** [comment]
Nice intro to counters in nidaq:
https://forums.ni.com/t5/Counter-Timer/Can-some-recommend-a-good-tutorials-on-how-to-use-timers/m-p/514941?requireLogin=False

**Comment by bruno-f-cruz** [comment]
Encoder datasheet:
https://www.distrelec.biz/Web/Downloads/t_/en/vt2400_210160-65_dat_en.pdf

**Comment by bruno-f-cruz** [comment]
After some testing, I came to the conclusion that the current NIDAQ board only supports digital IO in buffered mode on Port0. As a result, we can either use the buffered mode for digital inputs or outputs. I opted to use it for DigitalInputs since there were more pins and we are saving this input, so it is advantageous to log it in a more compressed format (i.e. 1 bit per channel as opposed to the full 16bit/channel format if using analog input). In order to be able to drive reward delivery, I moved the Reward trigger from the DO to AO. For now, it should work, however, it is worth noting that both AOs are currently taken (opto stimulation and reward).
